client/src/model/component/CircuitDiagram.hx

```diff
@@ -215,4 +215,124 @@ class CircuitDiagram extends Observable implements CircuitDiagramI implements Ob
         }
         return worldPointArray;
     }
+	
+	public function normalise() : Void {
+		// For all pairs of connections {c0, c1} such that c0 != c1 and c0 is close to c1
+		// if and both contain a port,
+			// create a link to connect them
+		// else one or both don't contain a port
+			// if c0 contains a port, swap c0 and c1
+			// move everything in c0 to c1
+		var connectionArray = new Array<Connection>();
+		for (link in this.linkArray) {
+			connectionArray.push(link.get_endpoint(0).getConnection());
+			connectionArray.push(link.get_endpoint(1).getConnection());
+		}
+		
+		for (component in this.componentArray) {
+			var ports = component.get_ports();
+			for (port in ports) {
+				connectionArray.push(port.getConnection());
+			}
+		}
+		
+		var markedConnections = new Array<Connection>();
+		for (connection in connectionArray) {
+			var otherConnections = connectionArray;
+			otherConnections.remove(connection);
+			for (otherConnection in otherConnections) {
+				var connectionDistance = Link.pointsDistance(connection.get_xPosition(), connection.get_yPosition(), otherConnection.get_xPosition(), otherConnection.get_yPosition());
+				if (connectionDistance <= 15) {
+					// here 15 is what is being considered as "close" to each other
+					if (connection.aPortIsConnecte() && otherConnection.aPortIsConnecte()) {
+						this.addLink(new Link(this, connection.get_xPosition(), connection.get_yPosition(), otherConnection.get_xPosition(), otherConnection.get_yPosition()));
+					} else {
+						if (connection.aPortIsConnecte()) {
```

Comment:
I knew this would not seem right, but, my doubt was, why swap, what do we achieve by keeping the swapped connection alive? The context of the use case seems to discard the connection off after it's constituent elements are pushed into the other connection.

Reply:
This is just a swap of the c0 and c1 pointers; no object modified by the swap; i.e. it is `{var temp = c0 ; c1 = c1 ; c1 = temp}`. The point of the swap is so that in the next step we move connectables from the one without the port to the one that has the port.  If we moved the connectables the other way then the port would move (unless c0 and c1 happen to be at exactly the same location) and I don't want normalization to move any ports.  Note that moving all connectables out of a connection will remove the connection from its circuit diagram and it will become garbage collectable.